Concurrency issue with listing uma policies yields HTTP 500

### Describe the bug

Keycloak throws HTTP 500 errors when querying policies during concurrent add/remove/query permission requests.  We encountered this on our internal deployment while running multiple applications using keycloak.  The steps here have been simplified down to what I believe are the minimum needed to generate the error.

### Version

18.0.1

### Expected behavior

No errors with valid concurrent requests

### Actual behavior

HTTP 500 w/ two server stack traces.

First stack trace:
14:21:21,625 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (default task-6) Uncaught server error: java.lang.RuntimeException: Could not create policy [uma] representation
        at org.keycloak.keycloak-server-spi-private@999-SNAPSHOT//org.keycloak.models.utils.ModelToRepresentation.toRepresentation(ModelToRepresentation.java:955)
        at org.keycloak.keycloak-services@999-SNAPSHOT//org.keycloak.authorization.admin.PermissionService.toRepresentation(PermissionService.java:64)
        at org.keycloak.keycloak-services@999-SNAPSHOT//org.keycloak.authorization.admin.PolicyService.lambda$doSearch$0(PolicyService.java:268)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.LinkedList$LLSpliterator.forEachRemaining(LinkedList.java:1239)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.keycloak.keycloak-services@999-SNAPSHOT//org.keycloak.authorization.admin.PolicyService.doSearch(PolicyService.java:269)
        at org.keycloak.keycloak-services@999-SNAPSHOT//org.keycloak.authorization.admin.PermissionService.doSearch(PermissionService.java:59)
        at org.keycloak.keycloak-services@999-SNAPSHOT//org.keycloak.authorization.admin.PolicyService.findAll(PolicyService.java:257)
        at org.keycloak.keycloak-services@999-SNAPSHOT//org.keycloak.authorization.protection.policy.UserManagedPermissionService.find(UserManagedPermissionService.java:130)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:249)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:60)
        at org.jboss.resteasy.resteasy-core@4.7.6.Final//org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:55)
        at javax.servlet.api@2.0.0.Final//javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
        at org.keycloak.keycloak-wildfly-extensions@999-SNAPSHOT//org.keycloak.provider.wildfly.WildFlyRequestFilter.lambda$doFilter$0(WildFlyRequestFilter.java:41)
        at org.keycloak.keycloak-services@999-SNAPSHOT//org.keycloak.services.filters.AbstractRequestFilter.filter(AbstractRequestFilter.java:43)
        at org.keycloak.keycloak-wildfly-extensions@999-SNAPSHOT//org.keycloak.provider.wildfly.WildFlyRequestFilter.doFilter(WildFlyRequestFilter.java:39)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
        at org.wildfly.security.elytron-web.undertow-server@1.10.1.Final//org.wildfly.elytron.web.undertow.server.ElytronRunAsHandler.lambda$handleRequest$1(ElytronRunAsHandler.java:68)
        at org.wildfly.security.elytron-base@1.19.0.Final//org.wildfly.security.auth.server.FlexibleIdentityAssociation.runAsFunctionEx(FlexibleIdentityAssociation.java:103)
        at org.wildfly.security.elytron-base@1.19.0.Final//org.wildfly.security.auth.server.Scoped.runAsFunctionEx(Scoped.java:161)
        at org.wildfly.security.elytron-base@1.19.0.Final//org.wildfly.security.auth.server.Scoped.runAs(Scoped.java:73)
        at org.wildfly.security.elytron-web.undertow-server@1.10.1.Final//org.wildfly.elytron.web.undertow.server.ElytronRunAsHandler.handleRequest(ElytronRunAsHandler.java:67)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:117)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
        at io.undertow.core@2.2.17.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at io.undertow.core@2.2.17.Final//io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
        at io.undertow.core@2.2.17.Final//io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
        at org.wildfly.security.elytron-web.undertow-server-servlet@1.10.1.Final//org.wildfly.elytron.web.undertow.server.servlet.CleanUpHandler.handleRequest(CleanUpHandler.java:38)
        at io.undertow.core@2.2.17.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at org.wildfly.extension.undertow@26.1.1.Final//org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
        at io.undertow.core@2.2.17.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at org.wildfly.extension.undertow@26.1.1.Final//org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.SendErrorPageHandler.handleRequest(SendErrorPageHandler.java:52)
        at io.undertow.core@2.2.17.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:275)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:79)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:134)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:131)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
        at org.wildfly.extension.undertow@26.1.1.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1544)
        at org.wildfly.extension.undertow@26.1.1.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1544)
        at org.wildfly.extension.undertow@26.1.1.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1544)
        at org.wildfly.extension.undertow@26.1.1.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1544)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:255)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:79)
        at io.undertow.servlet@2.2.17.Final//io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:100)
        at io.undertow.core@2.2.17.Final//io.undertow.server.Connectors.executeRootHandler(Connectors.java:387)
        at io.undertow.core@2.2.17.Final//io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:852)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
        at org.jboss.xnio@3.8.7.Final//org.xnio.XnioWorker$WorkerThreadFactory$1$1.run(XnioWorker.java:1282)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException
        at org.keycloak.keycloak-model-infinispan@999-SNAPSHOT//org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession.cachePolicy(StoreFactoryCacheSession.java:1232)
        at org.keycloak.keycloak-model-infinispan@999-SNAPSHOT//org.keycloak.models.cache.infinispan.authorization.PolicyAdapter.getAssociatedPolicies(PolicyAdapter.java:213)
        at org.keycloak.keycloak-authz-policy-common@999-SNAPSHOT//org.keycloak.authorization.policy.provider.permission.UMAPolicyProviderFactory.toRepresentation(UMAPolicyProviderFactory.java:317)
        at org.keycloak.keycloak-authz-policy-common@999-SNAPSHOT//org.keycloak.authorization.policy.provider.permission.UMAPolicyProviderFactory.toRepresentation(UMAPolicyProviderFactory.java:52)
        at org.keycloak.keycloak-server-spi-private@999-SNAPSHOT//org.keycloak.models.utils.ModelToRepresentation.toRepresentation(ModelToRepresentation.java:953)
        ... 93 more

Following stack trace(s):
2022-06-21 16:50:56,524 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-26) Uncaught server error: java.lang.RuntimeException: Could not create policy [uma] representation
        at org.keycloak.models.utils.ModelToRepresentation.toRepresentation(ModelToRepresentation.java:944)
        at org.keycloak.authorization.admin.PermissionService.toRepresentation(PermissionService.java:64)
        at org.keycloak.authorization.admin.PolicyService.lambda$doSearch$0(PolicyService.java:269)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.LinkedList$LLSpliterator.forEachRemaining(LinkedList.java:1239)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.keycloak.authorization.admin.PolicyService.doSearch(PolicyService.java:270)
        at org.keycloak.authorization.admin.PermissionService.doSearch(PermissionService.java:59)
        at org.keycloak.authorization.admin.PolicyService.findAll(PolicyService.java:258)
        at org.keycloak.authorization.protection.policy.UserManagedPermissionService.find(UserManagedPermissionService.java:131)
        at jdk.internal.reflect.GeneratedMethodAccessor395.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:201)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:67)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:55)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:201)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:380)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:358)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:201)
        at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:71)
        at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.RuntimeException: Could not create policy [user] representation
        at org.keycloak.models.utils.ModelToRepresentation.toRepresentation(ModelToRepresentation.java:944)
        at org.keycloak.models.utils.ModelToRepresentation.toRepresentation(ModelToRepresentation.java:927)
        at org.keycloak.authorization.policy.provider.permission.UMAPolicyProviderFactory.toRepresentation(UMAPolicyProviderFactory.java:317)
        at org.keycloak.authorization.policy.provider.permission.UMAPolicyProviderFactory.toRepresentation(UMAPolicyProviderFactory.java:52)
        at org.keycloak.models.utils.ModelToRepresentation.toRepresentation(ModelToRepresentation.java:942)
        ... 65 more
Caused by: java.lang.IllegalArgumentException: argument "content" is null
        at com.fasterxml.jackson.databind.ObjectMapper._assertNotNull(ObjectMapper.java:4829)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3596)
        at org.keycloak.util.JsonSerialization.readValue(JsonSerialization.java:75)
        at org.keycloak.authorization.policy.provider.user.UserPolicyProviderFactory.toRepresentation(UserPolicyProviderFactory.java:74)
        at org.keycloak.authorization.policy.provider.user.UserPolicyProviderFactory.toRepresentation(UserPolicyProviderFactory.java:45)
        at org.keycloak.models.utils.ModelToRepresentation.toRepresentation(ModelToRepresentation.java:942)
        ... 69 more

### How to Reproduce?

I think I've reduced this to the minimum number of steps in order to reproduce this problem.  Removing any of the steps will not reproduce the issue.

Start the server, add a client 

Running the following block of requests in multiple threads will cause the error:

1. Get a client token via POST request to http://[host]:[port]/[path]/realms/[realm]/protocol/openid-connect/token
2. Create a unique user via POST request to http://[host]:[port]/[path]/admin/realms/[realm]/users
3. Create a unique resource in the test client via POST request to http://[host]:[port]/[path]/realms/[realm]/authz/protection/resource_set
4. Create a unique policy for that resource and user with at least one authorization scope via POST request to http://[host]:[port]/[path]/realms/[realm]/authz/protection/uma-policy/[resource-id]
5. Delete the policy via DELETE request to http://[host]:[port]/[path]/realms/[realm]/authz/protection/uma-policy/[policy-id]
6. Get a list of existing policies via GET to http://[host]:[port]/[path]/realms/[realm]/authz/protection/uma-policy?first=0&max=500

Step #6 is what will generate the error (again, only during concurrent requests).  Also, if you replace user in the above steps with a role then you will get a similar error but with the RolePolicyProviderFactory instead of UserPolicyProviderFactory.


### Anything else?

_No response_

--- a/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/client/ClientPolicyProviderFactory.java
+++ b/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/client/ClientPolicyProviderFactory.java
@@ -19,7 +19,7 @@ package org.keycloak.authorization.policy.provider.client;
 
 import java.io.IOException;
 import java.util.ArrayList;
-import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -67,7 +67,7 @@ public class ClientPolicyProviderFactory implements PolicyProviderFactory<Client
     @Override
     public ClientPolicyRepresentation toRepresentation(Policy policy, AuthorizationProvider authorization) {
         ClientPolicyRepresentation representation = new ClientPolicyRepresentation();
-        representation.setClients(new HashSet<>(Arrays.asList(getClients(policy))));
+        representation.setClients(getClients(policy));
         return representation;
     }
 
@@ -88,7 +88,7 @@ public class ClientPolicyProviderFactory implements PolicyProviderFactory<Client
 
     @Override
     public void onImport(Policy policy, PolicyRepresentation representation, AuthorizationProvider authorization) {
-        updateClients(policy, new HashSet<>(Arrays.asList(getClients(policy))), authorization);
+        updateClients(policy, getClients(policy), authorization);
     }
 
     @Override
@@ -193,17 +193,17 @@ public class ClientPolicyProviderFactory implements PolicyProviderFactory<Client
         }
     }
 
-    private String[] getClients(Policy policy) {
+    private Set<String> getClients(Policy policy) {
         String clients = policy.getConfig().get("clients");
 
         if (clients != null) {
             try {
-                return JsonSerialization.readValue(clients.getBytes(), String[].class);
+                return JsonSerialization.readValue(clients, Set.class);
             } catch (IOException e) {
                 throw new RuntimeException("Could not parse clients [" + clients + "] from policy config [" + policy.getName() + "].", e);
             }
         }
 
-        return new String[]{};
+        return Collections.emptySet();
     }
 }

--- a/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/clientscope/ClientScopePolicyProviderFactory.java
+++ b/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/clientscope/ClientScopePolicyProviderFactory.java
@@ -19,6 +19,7 @@ package org.keycloak.authorization.policy.provider.clientscope;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -151,9 +152,15 @@ public class ClientScopePolicyProviderFactory implements PolicyProviderFactory<C
         ClientScopePolicyRepresentation representation = new ClientScopePolicyRepresentation();
 
         try {
-            representation
-                .setClientScopes(new HashSet<>(Arrays.asList(JsonSerialization.readValue(policy.getConfig().get("clientScopes"),
-                    ClientScopePolicyRepresentation.ClientScopeDefinition[].class))));
+            String clientScopes = policy.getConfig().get("clientScopes");
+
+            if (clientScopes == null) {
+                representation.setClientScopes(Collections.emptySet());
+            } else {
+                representation
+                        .setClientScopes(new HashSet<>(Arrays.asList(JsonSerialization.readValue(clientScopes,
+                                ClientScopePolicyRepresentation.ClientScopeDefinition[].class))));
+            }
         } catch (IOException e) {
             throw new RuntimeException("Failed to deserialize client scopes", e);
         }

--- a/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/group/GroupPolicyProviderFactory.java
+++ b/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/group/GroupPolicyProviderFactory.java
@@ -19,6 +19,7 @@ package org.keycloak.authorization.policy.provider.group;
 
 import java.io.IOException;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -213,7 +214,13 @@ public class GroupPolicyProviderFactory implements PolicyProviderFactory<GroupPo
         policy.setConfig(config);
     }
 
-    private HashSet<GroupPolicyRepresentation.GroupDefinition> getGroupsDefinition(Map<String, String> config) throws IOException {
-        return new HashSet<>(Arrays.asList(JsonSerialization.readValue(config.get("groups"), GroupPolicyRepresentation.GroupDefinition[].class)));
+    private Set<GroupPolicyRepresentation.GroupDefinition> getGroupsDefinition(Map<String, String> config) throws IOException {
+        String groups = config.get("groups");
+
+        if (groups == null) {
+            return Collections.emptySet();
+        }
+
+        return new HashSet<>(Arrays.asList(JsonSerialization.readValue(groups, GroupPolicyRepresentation.GroupDefinition[].class)));
     }
 }

--- a/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/role/RolePolicyProviderFactory.java
+++ b/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/role/RolePolicyProviderFactory.java
@@ -40,6 +40,7 @@ import org.keycloak.util.JsonSerialization;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -78,7 +79,14 @@ public class RolePolicyProviderFactory implements PolicyProviderFactory<RolePoli
         RolePolicyRepresentation representation = new RolePolicyRepresentation();
 
         try {
-            representation.setRoles(new HashSet<>(Arrays.asList(JsonSerialization.readValue(policy.getConfig().get("roles"), RolePolicyRepresentation.RoleDefinition[].class))));
+            String roles = policy.getConfig().get("roles");
+
+            if (roles == null) {
+                representation.setRoles(Collections.emptySet());
+            } else {
+                representation.setRoles(new HashSet<>(
+                        Arrays.asList(JsonSerialization.readValue(roles, RolePolicyRepresentation.RoleDefinition[].class))));
+            }
         } catch (IOException cause) {
             throw new RuntimeException("Failed to deserialize roles", cause);
         }

--- a/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/user/UserPolicyProviderFactory.java
+++ b/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/user/UserPolicyProviderFactory.java
@@ -19,6 +19,7 @@
 package org.keycloak.authorization.policy.provider.user;
 
 import java.io.IOException;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
@@ -71,7 +72,13 @@ public class UserPolicyProviderFactory implements PolicyProviderFactory<UserPoli
         UserPolicyRepresentation representation = new UserPolicyRepresentation();
 
         try {
-            representation.setUsers(JsonSerialization.readValue(policy.getConfig().get("users"), Set.class));
+            String users = policy.getConfig().get("users");
+
+            if (users == null) {
+                representation.setUsers(Collections.emptySet());
+            } else {
+                representation.setUsers(JsonSerialization.readValue(users, Set.class));
+            }
         } catch (IOException cause) {
             throw new RuntimeException("Failed to deserialize roles", cause);
         }

--- a/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/authorization/PolicyAdapter.java
+++ b/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/authorization/PolicyAdapter.java
@@ -206,10 +206,14 @@ public class PolicyAdapter implements Policy, CachedModel<Policy> {
         }
         if (associatedPolicies != null) return associatedPolicies;
         associatedPolicies = new HashSet<>();
-        PolicyStore policyStore = cacheSession.getPolicyStore();
+        PolicyStore policyStore = cacheSession.getPolicyStoreDelegate();
         String resourceServerId = cached.getResourceServerId();
         for (String id : cached.getAssociatedPoliciesIds(modelSupplier)) {
             Policy policy = policyStore.findById(InfinispanCacheStoreFactoryProviderFactory.NULL_REALM, cacheSession.getResourceServerStore().findById(InfinispanCacheStoreFactoryProviderFactory.NULL_REALM, resourceServerId), id);
+            if (policy == null) {
+                // probably because the policy was removed
+                continue;
+            }
             cacheSession.cachePolicy(policy);
             associatedPolicies.add(policy);
         }

--- a/model/jpa/src/main/java/org/keycloak/authorization/jpa/store/JPAPolicyStore.java
+++ b/model/jpa/src/main/java/org/keycloak/authorization/jpa/store/JPAPolicyStore.java
@@ -199,8 +199,9 @@ public class JPAPolicyStore implements PolicyStore {
 
         List<String> result = paginateQuery(query, firstResult, maxResults).getResultList();
         List<Policy> list = new LinkedList<>();
+        PolicyStore policyStore = provider.getStoreFactory().getPolicyStore();
         for (String id : result) {
-            Policy policy = provider.getStoreFactory().getPolicyStore().findById(JPAAuthorizationStoreFactory.NULL_REALM, resourceServer, id);
+            Policy policy = policyStore.findById(JPAAuthorizationStoreFactory.NULL_REALM, resourceServer, id);
             if (Objects.nonNull(policy)) {
                 list.add(policy);
             }

--- a/model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java
+++ b/model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java
@@ -162,10 +162,8 @@ public class MapPolicyStore implements PolicyStore {
         }
 
         return tx.read(withCriteria(mcb).pagination(firstResult, maxResults, SearchableFields.NAME))
-            .map(MapPolicyEntity::getId)
-            // We need to go through cache
-            .map(id -> authorizationProvider.getStoreFactory().getPolicyStore().findById(realm, resourceServer, id))
-            .collect(Collectors.toList());
+                .map(entityToAdapterFunc(realm, resourceServer))
+                .collect(Collectors.toList());
     }
 
     private DefaultModelCriteria<Policy> filterEntryToDefaultModelCriteria(Map.Entry<Policy.FilterOption, String[]> entry) {

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java
@@ -22,6 +22,9 @@ import org.junit.Assert;
 import org.keycloak.Config.Scope;
 import org.keycloak.authorization.AuthorizationSpi;
 import org.keycloak.authorization.DefaultAuthorizationProviderFactory;
+import org.keycloak.authorization.policy.provider.PolicyProviderFactory;
+import org.keycloak.authorization.policy.provider.PolicySpi;
+import org.keycloak.authorization.policy.provider.permission.UMAPolicyProviderFactory;
 import org.keycloak.authorization.store.StoreFactorySpi;
 import org.keycloak.cluster.ClusterSpi;
 import org.keycloak.common.util.Time;
@@ -213,6 +216,7 @@ public abstract class KeycloakModelTest {
 
     private static final Set<Class<? extends Spi>> ALLOWED_SPIS = ImmutableSet.<Class<? extends Spi>>builder()
       .add(AuthorizationSpi.class)
+      .add(PolicySpi.class)
       .add(ClientScopeSpi.class)
       .add(ClientSpi.class)
       .add(ComponentFactorySpi.class)
@@ -235,6 +239,7 @@ public abstract class KeycloakModelTest {
     private static final Set<Class<? extends ProviderFactory>> ALLOWED_FACTORIES = ImmutableSet.<Class<? extends ProviderFactory>>builder()
       .add(ComponentFactoryProviderFactory.class)
       .add(DefaultAuthorizationProviderFactory.class)
+      .add(PolicyProviderFactory.class)
       .add(DefaultExecutorsProviderFactory.class)
       .add(DeploymentStateProviderFactory.class)
       .add(DatastoreProviderFactory.class)

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/authz/ConcurrentAuthzTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/authz/ConcurrentAuthzTest.java
@@ -1,0 +1,188 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.testsuite.model.authz;
+
+import org.junit.Test;
+import org.keycloak.authorization.AuthorizationProvider;
+import org.keycloak.authorization.model.Policy;
+import org.keycloak.authorization.model.ResourceServer;
+import org.keycloak.authorization.store.StoreFactory;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.ClientProvider;
+import org.keycloak.models.Constants;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RealmProvider;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.models.utils.ModelToRepresentation;
+import org.keycloak.representations.idm.authorization.UmaPermissionRepresentation;
+import org.keycloak.representations.idm.authorization.UserPolicyRepresentation;
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+
+@RequireProvider(AuthorizationProvider.class)
+@RequireProvider(RealmProvider.class)
+@RequireProvider(ClientProvider.class)
+public class ConcurrentAuthzTest extends KeycloakModelTest {
+
+    private String realmId;
+    private String resourceServerId;
+    private String resourceId;
+    private String adminId;
+
+    @Override
+    protected void createEnvironment(KeycloakSession s) {
+        RealmModel realm = s.realms().createRealm("test");
+        realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
+
+        realmId = realm.getId();
+
+        ClientModel client = s.clients().addClient(realm, "my-server");
+
+        AuthorizationProvider authorization = s.getProvider(AuthorizationProvider.class);
+        StoreFactory aStore = authorization.getStoreFactory();
+
+        ResourceServer rs = aStore.getResourceServerStore().create(client);
+        resourceServerId = rs.getId();
+        resourceId =  aStore.getResourceStore().create(rs, "myResource", client.getClientId()).getId();
+        aStore.getScopeStore().create(rs, "read");
+
+        adminId = s.users().addUser(realm, "admin").getId();
+    }
+
+    @Override
+    protected void cleanEnvironment(KeycloakSession s) {
+        s.realms().removeRealm(realmId);
+    }
+
+    @Override
+    protected boolean isUseSameKeycloakSessionFactoryForAllThreads() {
+        return true;
+    }
+
+    @Test
+    public void testPermissionRemoved() {
+        IntStream.range(0, 500).parallel().forEach(index -> {
+            String permissionId = withRealm(realmId, (session, realm) -> {
+                AuthorizationProvider authorization = session.getProvider(AuthorizationProvider.class);
+                StoreFactory aStore = authorization.getStoreFactory();
+                ResourceServer rs = aStore.getResourceServerStore().findById(realm, resourceServerId);
+
+                UserModel u = session.users().addUser(realm, "user" + index);
+
+                UmaPermissionRepresentation permission = new UmaPermissionRepresentation();
+                permission.setName(KeycloakModelUtils.generateId());
+                permission.addUser(u.getUsername());
+                permission.addScope("read");
+
+                permission.addResource(resourceId);
+                permission.setOwner(adminId);
+                return aStore.getPolicyStore().create(rs, permission).getId();
+            });
+
+            withRealm(realmId, (session, realm) -> {
+                AuthorizationProvider authorization = session.getProvider(AuthorizationProvider.class);
+                StoreFactory aStore = authorization.getStoreFactory();
+
+                aStore.getPolicyStore().delete(realm, permissionId);
+                return null;
+            });
+
+            withRealm(realmId, (session, realm) -> {
+                AuthorizationProvider authorization = session.getProvider(AuthorizationProvider.class);
+                StoreFactory aStore = authorization.getStoreFactory();
+                ResourceServer rs = aStore.getResourceServerStore().findById(realm, resourceServerId);
+
+                Map<Policy.FilterOption, String[]> searchMap = new HashMap<>();
+                searchMap.put(Policy.FilterOption.TYPE, new String[]{"uma"});
+                searchMap.put(Policy.FilterOption.OWNER, new String[]{adminId});
+                searchMap.put(Policy.FilterOption.PERMISSION, new String[] {"true"});
+                Set<String> s = aStore.getPolicyStore().find(realm, rs, searchMap, 0, 500).stream().map(Policy::getId).collect(Collectors.toSet());
+                assertThat(s, not(contains(permissionId)));
+                return null;
+            });
+        });
+    }
+
+    @Test
+    public void testStaleCacheConcurrent() {
+        String permissionId = withRealm(realmId, (session, realm) -> {
+            AuthorizationProvider authorization = session.getProvider(AuthorizationProvider.class);
+            StoreFactory aStore = authorization.getStoreFactory();
+            UserModel u = session.users().getUserById(realm, adminId);
+            ResourceServer rs = aStore.getResourceServerStore().findById(realm, resourceServerId);
+
+
+            UmaPermissionRepresentation permission = new UmaPermissionRepresentation();
+            permission.setName(KeycloakModelUtils.generateId());
+            permission.addUser(u.getUsername());
+            permission.addScope("read");
+
+            permission.addResource(resourceId);
+            permission.setOwner(adminId);
+            return aStore.getPolicyStore().create(rs, permission).getId();
+        });
+
+        IntStream.range(0, 500).parallel().forEach(index -> {
+            String createdPolicyId = withRealm(realmId, (session, realm) -> {
+                AuthorizationProvider authorization = session.getProvider(AuthorizationProvider.class);
+                StoreFactory aStore = authorization.getStoreFactory();
+                ResourceServer rs = aStore.getResourceServerStore().findById(realm, resourceServerId);
+                Policy permission = aStore.getPolicyStore().findById(realm, rs, permissionId);
+
+                UserPolicyRepresentation userRep = new UserPolicyRepresentation();
+                userRep.setName("isAdminUser" + index);
+                userRep.addUser("admin");
+                Policy associatedPolicy = aStore.getPolicyStore().create(rs, userRep);
+                permission.addAssociatedPolicy(associatedPolicy);
+                return associatedPolicy.getId();
+            });
+
+            withRealm(realmId, (session, realm) -> {
+                AuthorizationProvider authorization = session.getProvider(AuthorizationProvider.class);
+                StoreFactory aStore = authorization.getStoreFactory();
+                ResourceServer rs = aStore.getResourceServerStore().findById(realm, resourceServerId);
+                Policy permission = aStore.getPolicyStore().findById(realm, rs, permissionId);
+
+                assertThat(permission.getAssociatedPolicies(), not(contains(nullValue())));
+                ModelToRepresentation.toRepresentation(permission, authorization);
+
+                return null;
+            });
+
+            withRealm(realmId, (session, realm) -> {
+                AuthorizationProvider authorization = session.getProvider(AuthorizationProvider.class);
+                StoreFactory aStore = authorization.getStoreFactory();
+                aStore.getPolicyStore().delete(realm, createdPolicyId);
+                return null;
+            });
+        });
+    }
+}

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/Infinispan.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/Infinispan.java
@@ -24,6 +24,8 @@ import org.keycloak.keys.infinispan.InfinispanCachePublicKeyProviderFactory;
 import org.keycloak.keys.infinispan.InfinispanPublicKeyStorageProviderFactory;
 import org.keycloak.models.ActionTokenStoreSpi;
 import org.keycloak.models.SingleUseObjectSpi;
+import org.keycloak.models.cache.authorization.CachedStoreFactorySpi;
+import org.keycloak.models.cache.infinispan.authorization.InfinispanCacheStoreFactoryProviderFactory;
 import org.keycloak.models.cache.CachePublicKeyProviderSpi;
 import org.keycloak.models.session.UserSessionPersisterSpi;
 import org.keycloak.models.sessions.infinispan.InfinispanActionTokenStoreProviderFactory;
@@ -61,6 +63,7 @@ public class Infinispan extends KeycloakModelParameters {
     static final Set<Class<? extends Spi>> ALLOWED_SPIS = ImmutableSet.<Class<? extends Spi>>builder()
       .add(AuthenticationSessionSpi.class)
       .add(CacheRealmProviderSpi.class)
+      .add(CachedStoreFactorySpi.class)
       .add(CacheUserProviderSpi.class)
       .add(InfinispanConnectionSpi.class)
       .add(StickySessionEncoderSpi.class)
@@ -77,6 +80,7 @@ public class Infinispan extends KeycloakModelParameters {
     static final Set<Class<? extends ProviderFactory>> ALLOWED_FACTORIES = ImmutableSet.<Class<? extends ProviderFactory>>builder()
       .add(InfinispanAuthenticationSessionProviderFactory.class)
       .add(InfinispanCacheRealmProviderFactory.class)
+      .add(InfinispanCacheStoreFactoryProviderFactory.class)
       .add(InfinispanClusterProviderFactory.class)
       .add(InfinispanConnectionProviderFactory.class)
       .add(InfinispanUserCacheProviderFactory.class)

